process/track/sovereignChainShardBlockTrack.go

```diff
@@ -50,3 +57,168 @@ func (scsbt *sovereignChainShardBlockTrack) ComputeLongestSelfChain() (data.Head
 func (scsbt *sovereignChainShardBlockTrack) GetSelfNotarizedHeader(_ uint32, offset uint64) (data.HeaderHandler, []byte, error) {
 	return scsbt.selfNotarizer.GetNotarizedHeader(scsbt.shardCoordinator.SelfId(), offset)
 }
+
+func (scsbt *sovereignChainShardBlockTrack) receivedHeader(headerHandler data.HeaderHandler, headerHash []byte) {
+	extendedShardHeader, isExtendedShardHeaderReceived := headerHandler.(*block.ShardHeaderExtended)
+	if isExtendedShardHeaderReceived {
+		scsbt.receivedExtendedShardHeader(extendedShardHeader, headerHash)
+		return
+	}
+
+	scsbt.receivedShardHeader(headerHandler, headerHash)
+}
+
+func (scsbt *sovereignChainShardBlockTrack) receivedExtendedShardHeader(
+	extendedShardHeaderHandler data.ShardHeaderExtendedHandler,
+	extendedShardHeaderHash []byte,
+) {
+
+	log.Debug("received extended shard header from network in block tracker",
+		"shard", extendedShardHeaderHandler.GetShardID(),
+		"epoch", extendedShardHeaderHandler.GetEpoch(),
+		"round", extendedShardHeaderHandler.GetRound(),
+		"nonce", extendedShardHeaderHandler.GetNonce(),
+		"hash", extendedShardHeaderHash,
+	)
+
+	if !scsbt.shouldAddExtendedShardHeader(extendedShardHeaderHandler) {
+		log.Trace("received extended shard header is out of range", "nonce", extendedShardHeaderHandler.GetNonce())
+		return
+	}
+
+	if !scsbt.addHeader(extendedShardHeaderHandler, extendedShardHeaderHash, core.SovereignChainShardId) {
+		log.Trace("received extended shard header was not added", "nonce", extendedShardHeaderHandler.GetNonce())
+		return
+	}
+
+	scsbt.doWhitelistWithExtendedShardHeaderIfNeeded(extendedShardHeaderHandler)
+	scsbt.blockProcessor.ProcessReceivedHeader(extendedShardHeaderHandler)
+}
+
+func (scsbt *sovereignChainShardBlockTrack) shouldAddExtendedShardHeader(extendedShardHeaderHandler data.ShardHeaderExtendedHandler) bool {
+	lastNotarizedHeader, _, err := scsbt.crossNotarizer.GetLastNotarizedHeader(core.SovereignChainShardId)
+
+	isFirstCrossNotarizedHeader := err != nil && errors.Is(err, process.ErrNotarizedHeadersSliceForShardIsNil)
+	if isFirstCrossNotarizedHeader {
+		return true
+	}
+
+	if err != nil {
+		log.Debug("shouldAddExtendedShardHeader.GetLastNotarizedHeader",
+			"shard", extendedShardHeaderHandler.GetShardID(),
+			"error", err.Error())
+		return false
+	}
+
+	lastNotarizedHeaderNonce := lastNotarizedHeader.GetNonce()
+
+	isHeaderOutOfRange := extendedShardHeaderHandler.GetNonce() > lastNotarizedHeaderNonce+uint64(scsbt.maxNumHeadersToKeepPerShard)
+	return !isHeaderOutOfRange
+}
+
+func (scsbt *sovereignChainShardBlockTrack) doWhitelistWithExtendedShardHeaderIfNeeded(extendedShardHeaderHandler data.ShardHeaderExtendedHandler) {
+	if check.IfNil(extendedShardHeaderHandler) {
+		return
+	}
+	if scsbt.isExtendedShardHeaderOutOfRange(extendedShardHeaderHandler) {
+		return
+	}
+
+	miniBlockHandlers := extendedShardHeaderHandler.GetIncomingMiniBlockHandlers()
+
+	keys := make([][]byte, 0)
```

Comment:
Does L129-L140 is common code? I think all block trackers instances do the whitelisting in the same way. Can we extract it in base?

Reply:
+1

Reply:
Not really the same. In shard header and metablock instances we are whitelisting the miniblocks hashes and not the tx hashes. But because for the simplicity, in sovereign, we already have incoming miniblocks included in extended shard headers received, we need to whitelist only the txs now. For whitelisting the txs, block tracker uses another component implemented in miniBlockTrack.go but this component is useless for sovereign, as the miniblocks wil not be sent and received in the "traditional" way. So, we will not have any received miniblock that will notify this mini block tracker component which is hooked to the miniblock's pool.

Reply:
ok